### DISCUSSION TITLE
Allow Hydration Based on a Prop

### DISCUSSION THREAD

**Discussion by amxmln** [withastro/roadmap#497] [Proposal]
# Summary

Allow partial hydration of framework components by passing a conditional `client:*` prop.

# Background & Motivation

Sometimes, especially when working with modular Astro pages, it would make sense to be able to conditionally hydrate a framework component. For example, it may only make sense to hydrate a Gallery component if there is more than one image in it.

In a recent project, I built an Image component that could optionally also take a video URL. When loaded, the component checked if the user had allowed video playback before. If not, it would prompt the player. If they had disallowed, it would show only the image as a fallback, otherwise, it would load the player and start the video.

Unfortunately, with the current `client:*` props, I either have to hydrate all of these Image components even if they don't have a video URL, or use a JSX conditional to include the same component twice, which bloats the code.

# Goals 

- Prevent duplicated code
- Allow more flexibility regarding partial hydration

# Example

The new `client:*` prop could look like this: `client:if={predicate}`, where `predicate` is a function or expression that resolves to a truthy or falsey value. When the value is truthy, the component will be hydrated. Otherwise it won't be hydrated.

**Comment by markpinero** [comment]
There is already a proposal for this #272

**Reply by amxmln** [reply]
I'm sorry, I must've missed that because I was looking for conditional client directives in particular. 😅
